Chem snap

charlessievers opened the pull request:
**Summary**

_Add chemical specificity to the current SNAP._

**Author(s)**

_Aidan Thompson Sandia National Labs
Charles Sievers UC Davis._

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

_All Current examples are runnable and produce the same output with no modifications to input._

**Implementation Notes**

_All Current examples are runnable and produce the same output with no modifications to input._

charlessievers commented:
> This needs modifications to the documentation to document the new options.

Will do.

charlessievers commented:
Hello @akohlmey,

I tried reading the console output for the checks I failed. Usually the console output is quite helpful, unfortunately this time it is not.

I would assume that these tests should have no dependency on the example files I added, therefore I am unsure why my last commit passed the checks and this commit failed.

Did you recently change the tests?

Best,
Charlie

akohlmey commented:
> Hello @akohlmey,
> 
> I tried reading the console output for the checks I failed. Usually the console output is quite helpful, unfortunately this time it is not.

Yup. C++ template programming can be "fun". ;-)

> I would assume that these tests should have no dependency on the example files I added, therefore I am unsure why my last commit passed the checks and this commit failed.

They failed because we updated the Kokkos library version and that required some changes
in how you have to specify compilation options. This has since been resolved in the test suite,
but you need to trigger another round of tests by pushing new commits, e.g. when you update
the documentation. ;-)

Please note, that your code also fails the coding style check: https://ci.lammps.org/job/lammps/job/pull-requests/job/checkstyle/806/
We just don't enforce that test, yet, but would appreciate if you address those as well.

> 
> Did you recently change the tests?
> 
> Best,
> Charlie

charlessievers commented:
> > Hello @akohlmey,
> > I tried reading the console output for the checks I failed. Usually the console output is quite helpful, unfortunately this time it is not.
> 
> Yup. C++ template programming can be "fun". ;-)
> 
> > I would assume that these tests should have no dependency on the example files I added, therefore I am unsure why my last commit passed the checks and this commit failed.
> 
> They failed because we updated the Kokkos library version and that required some changes
> in how you have to specify compilation options. This has since been resolved in the test suite,
> but you need to trigger another round of tests by pushing new commits, e.g. when you update
> the documentation. ;-)
> 
> Please note, that your code also fails the coding style check: https://ci.lammps.org/job/lammps/job/pull-requests/job/checkstyle/806/
> We just don't enforce that test, yet, but would appreciate if you address those as well.
> 
> > Did you recently change the tests?
> > Best,
> > Charlie

charlessievers commented:
> Yup. C++ template programming can be "fun". ;-)

I was actually trying a hand at some template programming in the sna.* files. I was hoping to make the SNA class similar to the post_force_templated() method in fix_langevin.cpp. There are several if statements deep in nested for loops within multiple methods. It would be nice if we could create an instance of SNA by SNA<some_flag>(arguments) and therefore not have to evaluate each if statement at run time. Is this possible? (Maybe I misunderstood the purpose of post_force_templated in fix_langevin.)

As always, thank you for your time. I appreciate your help.

> when you update the documentation. ;-)

Aidan and I will have some documentation soon.

akohlmey commented:
> > Yup. C++ template programming can be "fun". ;-)
> 
> I was actually trying a hand at some template programming in the sna.* files. I was hoping to make the SNA class similar to the post_force_templated() method in fix_langevin.cpp. There are several if statements deep in nested for loops within multiple methods. It would be nice if we could create an instance of SNA by SNA<some_flag>(arguments) and therefore not have to evaluate each if statement at run time. Is this possible? (Maybe I misunderstood the purpose of post_force_templated in fix_langevin.)

Yes. It is simpler to look this up in a lecture (pages 11 to 15) I gave at an HPC school at ICTP-SAIFR in Sao Paolo December 2019:
https://github.com/advinstai/hpc-school2019/blob/master/day5/optimizing-apps.pdf
The corresponding source code is here:
https://github.com/advinstai/hpc-school2019/tree/master/day5/lammps-opt

How much this helps depends on how "tight" the loop is.

charlessievers commented:
> Yes. It is simpler to look this up in a lecture (pages 11 to 15) I gave at an HPC school at ICTP-SAIFR in Sao Paolo December 2019:
> https://github.com/advinstai/hpc-school2019/blob/master/day5/optimizing-apps.pdf
> The corresponding source code is here:
> https://github.com/advinstai/hpc-school2019/tree/master/day5/lammps-opt
> 
> How much this helps depends on how "tight" the loop is.

Thank you, is there an example where the pair class itself is templated with an integer? I would like all methods in my class to be differentiated at compile time and without templating them all individually. (I got individual templated methods to work and saw a speed up a few days ago, but templating all classes individually would be far too cluttered to be worth it.)

I have some idea of what to do, but I keep getting linking errors when I call the templated classes methods from other .cpp files. If you are feeling generous with your time I could send you my failed prototype, if not I appreciate the help you have given me so far.

Great slides by the way. If you have more lectures on HPC, please send them my way. The coordinate structs are interesting.

akohlmey commented:
> Thank you, is there an example where the pair class itself is templated with an integer? I would like all methods in my class to be differentiated at compile time and without templating them all individually. (I got individual templated methods to work and saw a speed up a few days ago, but templating all classes individually would be far too cluttered to be worth it.)

The KOKKOS package does that. See for example: https://github.com/lammps/lammps/blob/master/src/KOKKOS/pair_lj_cut_kokkos.h

but that requires to associate each variant with a different style name, or have a "forwarder" class that will then create the specific class instances internally and then forward all methods to the internal instance.

> I have some idea of what to do, but I keep getting linking errors when I call the templated classes methods from other .cpp files. If you are feeling generous with your time I could send you my failed prototype, if not I appreciate the help you have given me so far.

I am too busy with several large and ambitious projects right now, taking advantage of having far fewer distractions when working from home: see #1986 #1981 #1970 #1842 and a couple more that are not posted yet.

> Great slides by the way. If you have more lectures on HPC, please send them my way. The coordinate structs are interesting.

My slides from ICTP events are usually posted with the corresponding events at https://www.ictp.it/scientific-calendar.aspx. I am lecturing at and co-organizing HPC events supported by ICTP for more than 15 years now. The repository I referred to was an event organized by the south american ICTP "franchise" has all the material and slides from that specific event (not just from me). There are some (older) slides posted on my homepage (which is notoriously outdated). For others you would have to attend my classes at the MHPC program in Italy https://www.mhpc.it/ as those are mostly given using a blackboard for didactic reasons.
I strongly believe that HPC is more a craft than a science, so text books and talk slides are of limited value and doing hands-on work is most important. My talk slides are usually just providing the list of talking points.

charlessievers commented:
> The KOKKOS package does that. See for example: https://github.com/lammps/lammps/blob/master/src/KOKKOS/pair_lj_cut_kokkos.h
> 
> but that requires to associate each variant with a different style name, or have a "forwarder" class that will then create the specific class instances internally and then forward all methods to the internal instance.

Thank you! My set up had a forwarder class as well, I must have just been missing a few things that these examples will help me with.


> I am too busy with several large and ambitious projects right now, taking advantage of having far fewer distractions when working from home: see #1986 #1981 #1970 #1842 and a couple more that are not posted yet.

I figured that would be the case.


> My slides from ICTP events are usually posted with the corresponding events at https://www.ictp.it/scientific-calendar.aspx. I am lecturing at and co-organizing HPC events supported by ICTP for more than 15 years now. The repository I referred to was an event organized by the south american ICTP "franchise" has all the material and slides from that specific event (not just from me). There are some (older) slides posted on my homepage (which is notoriously outdated). For others you would have to attend my classes at the MHPC program in Italy https://www.mhpc.it/ as those are mostly given using a blackboard for didactic reasons.
> I strongly believe that HPC is more a craft than a science, so text books and talk slides are of limited value and doing hands-on work is most important. My talk slides are usually just providing the list of talking points.

It is definitely a craft, which is why talking points such as that are useful. I enjoy clear concise examples of creative ways to optimize software.

Have a wonderful Easter and thank you again.

charlessievers commented:
@akohlmey New options have documented, examples have been added, and Kokkos has been updated.

stanmoore1 commented:
@athomps I do see a small performance regression for 2J14 on Summit:

Master:
```
log_d29n10_CPU:Loop time of 24.2442 on 36 procs for 100 steps with 2000 atoms
log_d29n10_GPU_1:Loop time of 4.79354 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_2:Loop time of 4.80361 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_3:Loop time of 4.79719 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_4:Loop time of 4.79369 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_5:Loop time of 4.77166 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_6:Loop time of 4.79771 on 1 procs for 200 steps with 2000 atoms
```

This PR:
```
log_d29n10_CPU:Loop time of 24.7802 on 36 procs for 100 steps with 2000 atoms
log_d29n10_GPU_1:Loop time of 5.04559 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_2:Loop time of 5.00856 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_3:Loop time of 4.99483 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_4:Loop time of 5.01217 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_5:Loop time of 5.00952 on 1 procs for 200 steps with 2000 atoms
log_d29n10_GPU_6:Loop time of 5.01474 on 1 procs for 200 steps with 2000 atoms
```

stanmoore1 commented:
I confirmed that the numerics for 2J14 on the GPU are correct though.

stanmoore1 commented on src/KOKKOS/sna_kokkos_impl.h:
```diff
@@ -223,13 +234,14 @@ void SNAKokkos<DeviceType>::grow_rij(int newnatom, int newnmax)
+  blist = t_sna_2d_ll("sna:blist",idxb_max*ntriples,natom);
   //ulisttot = t_sna_2c("sna:ulisttot",natom,idxu_max);
-  ulisttot = t_sna_2c_ll("sna:ulisttot",idxu_max,natom);
+  ulisttot = t_sna_2c_ll("sna:ulisttot",idxu_max*nelements,natom);
```
This may not give optimal memory access since the flattening loop assumes a memory layout which is the same for CPUs and GPUs. It may be better to make `elements` another dimension to the array so Kokkos can change the layout, but this could also hurt performance when `nelements=1` as in the ECP benchmarks.

stanmoore1 replied:
So I'm fine with leaving it the way it is.

athomps replied:
Agreed.

charlessievers replied:
We can try out adding element information to idxu in the future. We definitely will not need an extra dimension.

stanmoore1 replied:
OK I'm marking this point as resolved.

weinbe2 replied:
From an instruction count standpoint: Since all the Views are dynamically sized as opposed to statically sized, I believe making `elements` another dimension vs the flattened index is a performance wash---it's another fused integer multiply add, whether or not it's explicit in the functors.

Whether or not `idxu` should be faster than `nelements` (and the appropriate generalizations for `zlist`, `blist`, `ylist`) _in general_ is an interesting question related to cache reuse. For `nelements`, etc, `==1` that's a wash.

stanmoore1 commented on src/KOKKOS/sna_kokkos_impl.h:
```diff
@@ -869,67 +921,90 @@ void SNAKokkos<DeviceType>::compute_bi(const typename Kokkos::TeamPolicy<DeviceT
+      jalloy = idouble*idxz_max;
+      for (int elem3 = 0; elem3 < nelements; elem3++) {
+        Kokkos::parallel_for(Kokkos::TeamThreadRange(team, idxb_max),
+                             [&](const int &jjb) {
```
Can we reduce the indent level here?

athomps replied:
@charlessievers can you fix this and any other whitespace issues?

charlessievers replied:
Just to be clear, all you want is for me to reduce the whitespace in the for loops I created?

athomps replied:
Yes.

stanmoore1 replied:
Yeah, if the indent is too large then it makes it hard to read.

charlessievers replied:
@stanmoore1 Have a look at the new version and tell me if this is what you were thinking?

athomps commented:
@stanmoore1 thanks for checking te V100 performance. So it is about 5% slower.  I think this is acceptable, although it would good to know where the time being lost.

stanmoore1 commented:
@weinbe2 FYI, this PR is probably going to collide with your upcoming changes.

athomps commented:
@stanmoore1 we did not see a slowdown in the non-Kokkos timings.

weinbe2 commented:
@stanmoore1 oh most definitely :)

stanmoore1 commented:
Here is a profile using Kokkos tools. Looks like `ComputeYi` and `ComputeFusedDeidrj` both got a little slower:

Master:
```
<average time> <percent of total time> <percent time in Kokkos> <percent MPI imbalance> <number of calls> <name> [type]
|-> 2.91e+00 sec 58.1% 100.0% 0.0% ------ 201 ComputeYi [for]
|-> 1.10e+00 sec 22.0% 100.0% 0.0% ------ 603 ComputeFusedDeidrj [for]
|-> 6.74e-01 sec 13.5% 100.0% 0.0% ------ 201 ComputeUi [for]
|-> 3.43e-02 sec 0.7% 100.0% 0.0% ------ 2 ComputeZi [for]

KOKKOS CUDA SPACE:
=================== 
MAX MEMORY ALLOCATED: 893754.7 kB
ALLOCATIONS AT TIME OF HIGH WATER MARK:
  88.5% sna:zlist
  4.3% sna:ulisttot
  4.3% sna:ylist
  0.5% neighlist:neighbors
```



This PR:
```
<average time> <percent of total time> <percent time in Kokkos> <percent MPI imbalance> <number of calls> <name> [type]
=================== 
|-> 3.00e+00 sec 57.7% 100.0% 0.0% ------ 201 ComputeYi [for]
|-> 1.20e+00 sec 23.1% 100.0% 0.0% ------ 603 ComputeFusedDeidrj [for]
|-> 6.71e-01 sec 12.9% 100.0% 0.0% ------ 201 ComputeUi [for]
|-> 3.44e-02 sec 0.7% 100.0% 0.0% ------ 2 ComputeZi [for]

KOKKOS CUDA SPACE:
=================== 
MAX MEMORY ALLOCATED: 894207.8 kB
ALLOCATIONS AT TIME OF HIGH WATER MARK:
  88.4% sna:zlist
  4.3% sna:ulisttot
  4.3% sna:ylist
  0.5% neighlist:neighbors
```

charlessievers commented:
> Here is a profile using Kokkos tools. Looks like `ComputeYi` and `ComputeFusedDeidrj` both got a little slower:

I have a low hanging that might speed up Deirdrj slightly.

stanmoore1 commented:
I'm amazed we are now using less than 1 GB of GPU memory for 2J14 with 2000 atoms. We could easily bump up the default chunksize to more than 2000.

weinbe2 commented on src/KOKKOS/sna_kokkos_impl.h:
```diff
@@ -532,79 +560,101 @@ KOKKOS_INLINE_FUNCTION
+  int itriple;
+  for(int elem1 = 0; elem1 < nelements; elem1++)
+    for (int elem2 = 0; elem2 < nelements; elem2++) {
+      double betaj;
```
You can move this line through the line reading `const double* cgblock = [...]` out of the loop over `elem1`, `elem2` to save extra loads for `nelements > 1`, since these are independent of the element.

weinbe2 replied:
Same in `compute_zi`. It's trickier in `compute_bi` because of the hierarchical parallelism.

charlessievers replied:
Thanks @weinbe2 ! I just implemented this.

charlessievers replied:
@stanmoore1 I also updated a few other portions that will most likely increase performance.

charlessievers commented:
> I'm amazed we are now using less than 1 GB of GPU memory for 2J14 with 2000 atoms. We could easily bump up the default chunksize to more than 2000.

This is most likely due the removal of gpu dulist initialization, if I had to imagine.

weinbe2 commented:
> > I'm amazed we are now using less than 1 GB of GPU memory for 2J14 with 2000 atoms. We could easily bump up the default chunksize to more than 2000.
> 
> This is most likely due the removal of gpu dulist initialization, if I had to imagine.

100% was, and considering I never shut up about figuring out how to remove it I still can't believe I accidentally left it in!

stanmoore1 commented:
Here is the performance with the updates from today. A tiny bit faster:

```
log_h50n02_CPU:Loop time of 24.566 on 36 procs for 100 steps with 2000 atoms
log_h50n02_GPU_1:Loop time of 4.97215 on 1 procs for 200 steps with 2000 atoms
log_h50n02_GPU_2:Loop time of 4.95233 on 1 procs for 200 steps with 2000 atoms
log_h50n02_GPU_3:Loop time of 4.97489 on 1 procs for 200 steps with 2000 atoms
log_h50n02_GPU_4:Loop time of 4.98503 on 1 procs for 200 steps with 2000 atoms
log_h50n02_GPU_5:Loop time of 4.95736 on 1 procs for 200 steps with 2000 atoms
log_h50n02_GPU_6:Loop time of 4.98658 on 1 procs for 200 steps with 2000 atoms
```

charlessievers commented:
I think extending idxu/idxz/idxb is the way to go to see the same performance as before.

weinbe2 commented:
I noticed there were no changes to the `TagPairSNAPComputeForce`-tagged routines related to chem_snap, even though the energy contribution depends on `blist` which has a new `ntriple` dependency. Is this correct? Or is there an error in the energy calculation for `ntriple != 1`?

charlessievers commented:
> I noticed there were no changes to the `TagPairSNAPComputeForce`-tagged routines related to chem_snap, even though the energy contribution depends on `blist` which has a new `ntriple` dependency. Is this correct? Or is there an error in the energy calculation for `ntriple != 1`?
> 
> Representative lines where `blist` is used (for Ctrl+F purposes):
> 
> ```
> for (int icoeff = 0; icoeff < ncoeff; icoeff++)
>   evdwl += d_coeffi[icoeff+1]*my_sna.blist(icoeff,ii);
> ```

I am rather confident that the answer to your question is here:

```
int SNAKokkos<DeviceType>::compute_ncoeff()
{
  int ncount;

  ncount = 0;

  for (int j1 = 0; j1 <= twojmax; j1++)
    for (int j2 = 0; j2 <= j1; j2++)
      for (int j = j1 - j2;
           j <= MIN(twojmax, j1 + j2); j += 2)
        if (j >= j1) ncount++;

  ndoubles = nelements*nelements;
  ntriples = nelements*nelements*nelements;
  if (chem_flag) ncount *= ntriples;

  return ncount;
```

weinbe2 commented:
> > I noticed there were no changes to the `TagPairSNAPComputeForce`-tagged routines related to chem_snap, even though the energy contribution depends on `blist` which has a new `ntriple` dependency. Is this correct? Or is there an error in the energy calculation for `ntriple != 1`?
> > Representative lines where `blist` is used (for Ctrl+F purposes):
> > ```
> > for (int icoeff = 0; icoeff < ncoeff; icoeff++)
> >   evdwl += d_coeffi[icoeff+1]*my_sna.blist(icoeff,ii);
> > ```
> 
> I am rather confident that the answer to your question is here:
> 
> ```
> int SNAKokkos<DeviceType>::compute_ncoeff()
> {
>   int ncount;
> 
>   ncount = 0;
> 
>   for (int j1 = 0; j1 <= twojmax; j1++)
>     for (int j2 = 0; j2 <= j1; j2++)
>       for (int j = j1 - j2;
>            j <= MIN(twojmax, j1 + j2); j += 2)
>         if (j >= j1) ncount++;
> 
>   ndoubles = nelements*nelements;
>   ntriples = nelements*nelements*nelements;
>   if (chem_flag) ncount *= ntriples;
> 
>   return ncount;
> ```

Ah, perfect, thank you.

weinbe2 commented:
FYI @charlessievers I merged your PR into the optimizations I'd been working on (specifically for `Compute[Yi/Zi/Bi]`).

The optimizations to `ComputeYi` more than amortize the small regressions you hit there, and I figured out a few tricks in `ComputeFusedDeidrj` to patch up the regression there. (It also has some nice GPU perf benefits for the chem snap example that gets added with this PR.)

charlessievers commented:
> FYI @charlessievers I merged your PR into the optimizations I'd been working on (specifically for `Compute[Yi/Zi/Bi]`).
> 
> The optimizations to `ComputeYi` more than amortize the small regressions you hit there, and I figured out a few tricks in `ComputeFusedDeidrj` to patch up the regression there. (It also has some nice GPU perf benefits for the chem snap example that gets added with this PR.)
> 
> Ideally I'll open my PR shortly after yours is merged.

That is great news! I am glad that they were not too much of a pain to merge.
